### PR TITLE
Multithreading support [ESD-1204]

### PR DESCRIPTION
**Pull request by pmiettinen**
Add support for parallel requests to boost up the performance with Swift Console. In single thread mode the round-trip-time sums to noticeable delay when req-resp pairs are done one by one in read/write all cases.

Retain single thread support for PBR and PFWP.

**TODO**
- [x] Update unit tests
- [x] Update wheels

**Used/Tested in**
https://github.com/swift-nav/piksi_tools/pull/938
https://github.com/swift-nav/piksi_buildroot/pull/1173
https://github.com/swift-nav/piksi_firmware_private/pull/2800

### CODE REVIEW

**Comment by benjaminaltieri** [comment]
this is a great POC and seems to work already for the console solution. will the `pthreads` dependency cause issues for PFWP/PBR integration? maybe there is a quick an dirty way of disabling threading on those platforms since I don't believe they need the same level of performance

**Comment by pmiettinen** [comment]
Actually `pthreads` is not used. The linker directive in the cmake file is leftover from my experiments. Current approach is that the client provides the locking and signaling mechanisms similar to earlier solution. Although the passing mechanisms probably need some tuning still.

**Comment by pmiettinen** [comment]
I put up a setup where I can test over serial instead of network and results look good.

Using `settings_client.py` (modified to do only `read_all`) `time` gives multi-threaded version 3-4 seconds real time results while single-thread is around 7-8 seconds. These times include all handler, serial_link etc constructing.

On console there's similar few second improvement. Compared to the original console indexed read I think there's still very minor performance drop but I'd say it's noticeable only if you know to look for it.

File: src/request_state.c [hunk 1]
```diff
@@ -111,3 +123,91 @@ void request_state_deinit(request_state_t *state)
 {
   state->pending = false;
 }
+
+void request_state_append(settings_t *ctx, request_state_t *state_data)
+{
+  if (ctx->client_iface.lock) {
```
Comment (pmiettinen) [inline src/request_state.c]: Could add a macro for these.

File: include/libsettings/settings.h [hunk 2]
```diff
@@ -463,6 +503,7 @@ LIBSETTINGS_DECLSPEC int settings_read_bool(settings_t *ctx,
  * @retval >0               Last index was read successfully. There are no more indexes to read.
  */
 LIBSETTINGS_DECLSPEC int settings_read_by_idx(settings_t *ctx,
+                                              void *event,
```
Comment (pmiettinen) [inline include/libsettings/settings.h]: add doc

**Comment by pmiettinen** [comment]
Latest CI Windows builds are giving error `CommandNotFoundError: No command 'conda conda'.`. Origin currently unknown. Investigating.

**Comment by pmiettinen** [comment]
Pinpointed to line `source activate py<version>` stopped working for some reason in the Travis environment.

**Comment by pmiettinen** [comment]
Fixed by reworking the sanity check in windows scripts.

File: python/settings_client.py [hunk 3]
```diff
@@ -88,13 +95,19 @@ def main():
 
             time.sleep(1)
 
+            slist = [{"section": "solution", "name": "correction_age_max", "value": "7"),
```
Comment (benjaminaltieri) [inline python/settings_client.py]: mismatched closing `)` to open `{`

Comment (pmiettinen) [reply python/settings_client.py]: Thanks, fixed and sanity checked.

**Pull request merged**